ballerina-to-openapi/src/main/java/io/ballerina/openapi/service/mapper/interceptor/Interceptor.java

```diff
@@ -0,0 +1,200 @@
+                    "interceptor class is defined within the package");
+        }
+        this.serviceClass = typeSymbol.typeDescriptor() instanceof ClassSymbol ?
+                (ClassSymbol) typeSymbol.typeDescriptor() : null;
+        if (Objects.isNull(this.serviceClass)) {
+            throw new InterceptorMapperException("no class definition found for the interceptor: " + name +
+                    ". Make sure that the interceptor return type is defined with the specific interceptor class " +
+                    "type rather than the generic `http:Interceptor` type");
+        }
+        extractInterceptorDetails(semanticModel);
+    }
+
+    protected abstract void extractInterceptorDetails(SemanticModel semanticModel);
+
+    public abstract boolean isInvokable(TargetResource targetResource);
+
+    protected void setReturnType(TypeSymbol returnType) {
+        hasNilReturn = semanticModel.types().NIL.subtypeOf(returnType);
+        TypeSymbol effectiveReturnType;
+        if (isSubTypeOfDefaultInterceptorReturnType(returnType, semanticModel)) {
+            continueExecution = true;
+            effectiveReturnType = getEffectiveReturnType(returnType, semanticModel);
+        } else {
+            effectiveReturnType = returnType;
+        }
+
+        if (Objects.isNull(effectiveReturnType)) {
+            return;
+        }
+
+        extractErrorAndNonErrorReturnTypes(effectiveReturnType);
+    }
+
+    private boolean isSubTypeOfDefaultInterceptorReturnType(TypeSymbol typeSymbol, SemanticModel semanticModel) {
+        Optional<Symbol> optNextServiceType = semanticModel.types().getTypeByName(BALLERINA, HTTP,
+                EMPTY, NEXT_SERVICE);
+        if (optNextServiceType.isEmpty() ||
+                !(optNextServiceType.get() instanceof TypeDefinitionSymbol nextServiceType)) {
+            return false;
+        }
+        UnionTypeSymbol defaultInterceptorReturnType = semanticModel.types().builder().UNION_TYPE.withMemberTypes(
+                nextServiceType.typeDescriptor(), semanticModel.types().NIL).build();
+        return defaultInterceptorReturnType.subtypeOf(typeSymbol);
+    }
+
+    private boolean isSubTypeOfHttpNextServiceType(TypeSymbol typeSymbol, SemanticModel semanticModel) {
+        Optional<Symbol> optNextServiceType = semanticModel.types().getTypeByName(BALLERINA, HTTP,
+                EMPTY, NEXT_SERVICE);
+        if (optNextServiceType.isEmpty() ||
+                !(optNextServiceType.get() instanceof TypeDefinitionSymbol nextServiceType)) {
+            return false;
+        }
+        return typeSymbol.subtypeOf(nextServiceType.typeDescriptor());
+    }
+
+    private TypeSymbol getEffectiveReturnType(TypeSymbol typeSymbol, SemanticModel semanticModel) {
+        if (isSubTypeOfHttpNextServiceType(typeSymbol, semanticModel) ||
+                typeSymbol.subtypeOf(semanticModel.types().NIL)) {
+            return null;
```

Comment:
What will be the effective return, if this return `null` ? will it fall back to the main service return?